### PR TITLE
Simple Payments: Convert the `multiple` value to boolean when working with API responses

### PR DESCRIPTION
**Pull request by jsnajdr**
Fixes the issue discussed in https://github.com/Automattic/wp-calypso/pull/16772#discussion_r130579768 where `multiple` is not a simple boolean, but a string (`""`, `"0"`, `"1"`, ...) and where JS and PHP have different interpretations of what is truthy and what is falsy.

This PR is divided into several commits:
- the first two are refactorings of `customPostToProduct` and `productToCustomPost` -- the `reduce` calls were quite challenging to understand, so I hope I made them a bit clear
- the real meat is in the following two commits, which introduce converting the `boolean` values and simplify the `SimplePaymentsView`

It took some effort, but the conversion functions should be more clear and less buggy now.

To test:
- edit a payment button and change the "Allow people to buy more than one item at a time" toggle
- check that the button preview in the editor shows or doesn't show the field for specifying the item count
- check that the frontend shows/doesn't show the input, too
- in short, do all the components both in Calypso and on frontend interpret `multiple` the same way?

### CODE REVIEW

File: client/state/data-layer/wpcom/sites/simple-payments/index.js [hunk 1]
```diff
@@ -28,33 +28,44 @@ import formatCurrency from 'lib/format-currency';
...
-	if ( metaKeyToSchemaKeyMap[ current.key ] ) {
-		sanitizedProductAttributes[ metaKeyToSchemaKeyMap[ current.key ] ] = current.value;
-	}
-	return sanitizedProductAttributes;
+function customPostMetadataToProductAttributes( metadata ) {
+	const productAttributes = {};
+
+	metadata.forEach( ( { key, value } ) => {
+		const schemaKey = metaKeyToSchemaKeyMap[ key ];
+		if ( schemaKey ) {
```
Comment (omarjackman) [inline client/state/data-layer/wpcom/sites/simple-payments/index.js]: If we return early we can get rid of the nesting `if` and indentation
```
if ( ! schemaKey ) {
    return;
}
```

File: client/state/data-layer/wpcom/sites/simple-payments/index.js [hunk 2]
```diff
@@ -28,33 +28,44 @@ import formatCurrency from 'lib/format-currency';
...
+			// If the property's type is marked as boolean in the schema,
+			// convert the value from PHP-ish truthy/falsy numbers to a plain boolean.
+			if ( metadataSchema[ schemaKey ].type === 'boolean' ) {
+				value = !! Number( value );
```
Comment (omarjackman) [inline client/state/data-layer/wpcom/sites/simple-payments/index.js]: Is the use of `Number` necessary here? Shouldn't `!! value` suffice? The only difference here I think would be `!! Number( 'true' )` and `!! Number( 'false' )` are always false.

File: client/state/data-layer/wpcom/sites/simple-payments/index.js [hunk 3]
```diff
@@ -63,35 +74,40 @@ export function customPostToProduct( product ) {
  * @returns { Object } custom post type data
  */
 export function productToCustomPost( product ) {
+	// Get the `product` object entries and filter only those that will go into metadata
+	const metadataEntries = Object.entries( product ).filter( ( [ key ] ) => metadataSchema[ key ] );
```
Comment (omarjackman) [inline client/state/data-layer/wpcom/sites/simple-payments/index.js]: `Object.entries` is [not supported](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#Browser_compatibility) in IE/Opera 🙁

Comment (jsnajdr) [inline client/state/data-layer/wpcom/sites/simple-payments/index.js]: Oh, these missing ES6 polyfills! I'll use `lodash/toPairs` instead then.

File: client/state/data-layer/wpcom/sites/simple-payments/index.js (hunk 2 above)
Comment (jsnajdr) [inline client/state/data-layer/wpcom/sites/simple-payments/index.js]: We need to treat `"0"` as falsy and `"1"` as truthy. Mere `!! value` would treat `"0"` as truthy though. I'll add a comment and also a unit test, as this conversion deserves some clarification.

**Comment by jsnajdr** [comment]
Pushed a few fixes:
- use `lodash/toPairs` instead of `Object.entries`
- add comment about the `!! Number( value )` conversion
- added unit test for the conversion functions

Ready for the next review round.
